### PR TITLE
Code Review - Write Path

### PR DESCRIPTION
**Pull request by arunoda**
[Review on Reviewable](https://reviewable.io/reviews/kadirahq/kadiyadb/1)

### CODE REVIEW

**Comment by arunoda** [comment]
**EDIT:**
use json 

Review status: 0 of 30 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[kadiyadb.go, line 363 [r1]](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1#-JzdLm6tPf4XPdRZtj6k)** ([raw file](https://github.com/kadirahq/kadiyadb/blob/d66e5f3b69787d9cd2e03289a6b06f6bebdeadfe/kadiyadb.go#L363)):</sup>
If we keep this duration and the resolution as unchangable metadata. We don't need this lock.  
Also  we need to make sure, we don't have any APIs to update this data.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1)

**Comment by arunoda** [comment]
Review status: 0 of 30 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[kadiyadb.go, line 616 [r1]](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1#-JzdMXskknzjR-ZumojM)** ([raw file](https://github.com/kadirahq/kadiyadb/blob/d66e5f3b69787d9cd2e03289a6b06f6bebdeadfe/kadiyadb.go#L616)):</sup>
We'll keep this MaxRWEpochs as a startTime configuration param. Then, we don't this local as well.

If we need to change this, we need to restart the DB. Which is fine.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1)

**Comment by arunoda** [comment]
Review status: 0 of 30 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

<sup>**[kadiyadb.go, line 712 [r1]](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1#-JzdO7nce3TtAXelJKYU)** ([raw file](https://github.com/kadirahq/kadiyadb/blob/d66e5f3b69787d9cd2e03289a6b06f6bebdeadfe/kadiyadb.go#L712)):</sup>
So, this store missing epochs. I assume this is for logging purpose.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1)

**Comment by arunoda** [comment]
Review status: 0 of 30 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[epoch.go, line 153 [r1]](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1#-JzdPR-oEe5HxoyIPqj-)** ([raw file](https://github.com/kadirahq/kadiyadb/blob/d66e5f3b69787d9cd2e03289a6b06f6bebdeadfe/epoch.go#L153)):</sup>
May be write long names like `recordId`. It's clear then when reading the code.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1)

**Comment by arunoda** [comment]
Review status: 0 of 30 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

<sup>**[epoch.go, line 154 [r1]](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1#-JzdQI_XHodfv9Ew5IMo)** ([raw file](https://github.com/kadirahq/kadiyadb/blob/d66e5f3b69787d9cd2e03289a6b06f6bebdeadfe/epoch.go#L154)):</sup>
I think this is the place where, multi level put should happen. Index needs to give us a list of record ids.

With the atomic increments, we could do this without any locks.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1)

**Comment by arunoda** [comment]
Review status: 0 of 30 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

<sup>**[epoch.go, line 158 [r1]](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1#-JzdQpFLpWYP-jSLQWxB)** ([raw file](https://github.com/kadirahq/kadiyadb/blob/d66e5f3b69787d9cd2e03289a6b06f6bebdeadfe/epoch.go#L158)):</sup>
Better if we can put this logic into a separate function. So, then it's easy to read.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1)

**Comment by arunoda** [comment]
Review status: 0 of 30 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

<sup>**[kadiyadb.go, line 354 [r1]](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1#-JzdRGUMoVa8XvWYCKqc)** ([raw file](https://github.com/kadirahq/kadiyadb/blob/d66e5f3b69787d9cd2e03289a6b06f6bebdeadfe/kadiyadb.go#L354)):</sup>
Here also, we get `value` and `samples` instead of a byte array?

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1)

**Comment by arunoda** [comment]
Reviewed 2 of 30 files at r1.
Review status: 2 of 30 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[block/block.go, line 264 [r1]](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1#-JzdRqXelTk0eFC4YRZz)** ([raw file](https://github.com/kadirahq/kadiyadb/blob/d66e5f3b69787d9cd2e03289a6b06f6bebdeadfe/block/block.go#L264)):</sup>
So, here we avoid these locks with atomic writes?

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1)

**Comment by arunoda** [comment]
Review status: 2 of 30 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

<sup>**[block/block.go, line 231 [r1]](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1#-JzdSiflAkgQZvyynjww)** ([raw file](https://github.com/kadirahq/kadiyadb/blob/d66e5f3b69787d9cd2e03289a6b06f6bebdeadfe/block/block.go#L231)):</sup>
Can we get rid of this lock as well? I assume with atomic updates that'll be okay too.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1)

**Comment by arunoda** [comment]
Reviewed 1 of 30 files at r1.
Review status: 3 of 30 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

<sup>**[block/block.go, line 264 [r1]](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1#-JzdRqXelTk0eFC4YRZz)** ([raw file](https://github.com/kadirahq/kadiyadb/blob/d66e5f3b69787d9cd2e03289a6b06f6bebdeadfe/block/block.go#L264)):</sup>
Here it might be possible to have slightly different values for the sample and value when reading. 

**But I assume, that's fine and our DB is eventually consistent always.**

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1)

**Comment by arunoda** [comment]
Review status: 3 of 30 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

<sup>**[index/index.go, line 350 [r1]](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1#-JzdV7Aq6vRdU7cA39Hy)** ([raw file](https://github.com/kadirahq/kadiyadb/blob/d66e5f3b69787d9cd2e03289a6b06f6bebdeadfe/index/index.go#L350)):</sup>
I think we don't need this lock.

---

<sup>**[index/index.go, line 511 [r1]](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1#-JzdVxXGT4lRilqSOJm6)** ([raw file](https://github.com/kadirahq/kadiyadb/blob/d66e5f3b69787d9cd2e03289a6b06f6bebdeadfe/index/index.go#L511)):</sup>
I hope we can use mmaped flat buffers here? Can't we?

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1)

**Comment by arunoda** [comment]
Review status: 3 of 30 files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

<sup>**[index/index.go, line 348 [r1]](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1#-JzdWfqxFtPkFuLQjkmQ)** ([raw file](https://github.com/kadirahq/kadiyadb/blob/d66e5f3b69787d9cd2e03289a6b06f6bebdeadfe/index/index.go#L348)):</sup>
I think this comment should goes with the above `i.store()` right?
I think we always lost some stuff, since we do the sync later. I think that's okay. 

If the server crashes before syncing, we might have some data in the block which has no index reference. I think that's not bad since it only leads to some hidden data.

We can later do some checks to see those kinds of data. May be reuse that space by detecting them in the start.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1)

**Comment by arunoda** [comment]
Review status: 3 of 30 files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

<sup>**[index/index.go, line 350 [r1]](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1#-JzdV7Aq6vRdU7cA39Hy)** ([raw file](https://github.com/kadirahq/kadiyadb/blob/d66e5f3b69787d9cd2e03289a6b06f6bebdeadfe/index/index.go#L350)):</sup>
I assume we **need** this lock. Ignore above comment. In the `epoch.Put` you have double checked for index exists and then continue?

So, I think you need to start this lock at the very beginning of this file. Otherwise, there could be possibilities to add the same index twice.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1)

**Comment by arunoda** [comment]
Review status: 3 of 30 files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

<sup>**[index/index.go, line 106 [r1]](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1#-Jzd_GnAbe_I_l9hD2U8)** ([raw file](https://github.com/kadirahq/kadiyadb/blob/d66e5f3b69787d9cd2e03289a6b06f6bebdeadfe/index/index.go#L106)):</sup>
I think this function should get all the parent nodes as well. May need to change a bit.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1)

**Comment by arunoda** [comment]
Review status: 3 of 30 files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

<sup>**[index/index.go, line 321 [r1]](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1#-Jzd_eBu8umFqR8xg5IY)** ([raw file](https://github.com/kadirahq/kadiyadb/blob/d66e5f3b69787d9cd2e03289a6b06f6bebdeadfe/index/index.go#L321)):</sup>
Here we need to get more than one value (recordIds)? I assume that's needed to store them middle of the tree, if there no existing records.

So, we may need to pass a closure as the value, which generate recordIds. Is that an option?

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1)

**Comment by arunoda** [comment]
Reviewed 1 of 30 files at r1.
Review status: 4 of 30 files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/kadirahq/kadiyadb/1)

File: epoch.go [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+		rid = item.Value
+	} else if goerr.Is(err, index.ErrNoItem) {
+		id, err := e.block.Add()
+		if err != nil {
+			return goerr.Wrap(err, 0)
+		}
+
+		err = e.index.Put(fields, id)
+		if err != nil && !goerr.Is(err, index.ErrExists) {
+			return goerr.Wrap(err, 0)
```
Comment (thani-sh) [inline epoch.go:165]: If record already exists, get the correct record id with a `One` request.

File: index/index.go [hunk 2]
```diff
@@ -0,0 +1,962 @@
...
+		children: make(map[string]*node),
+	}
+
+	if err = i.store(n); err != nil {
+		return goerr.Wrap(err, 0)
+	}
+
+	// index item should be saved before adding it to the in memory index
+	// otherwise index may miss some items when the server restarts
+	i.mutex.Lock()
```
Comment (thani-sh) [inline index/index.go:350]: Lock before checking on line 335

**Comment by thani-sh** [comment]
Sync memmaps only once per batch

**Pull request closed without merging**